packages/config/src/project/index.ts

```diff
@@ -0,0 +1,134 @@
+import _ from 'lodash'
+import Debug from 'debug'
+import deepDiff from 'return-deep-diff'
+
+import errors, { ConfigValidationFailureInfo, CypressError } from '@packages/errors'
+import { getCtx } from '@packages/data-context/src/globalContext'
```

Comment:
`getCtx` is called in 2 functions:

1. `updateWithPluginValues` of `project/index.ts`
2. `setSupportFileAndFolder` of `project/utils.ts`

The solution for the first case is simple. We can simply add `configFile` member inside `buildBaseFullConfig` of `ProjectConfigManager`.

But the problem is the second one. 

To call the `ctx.file.getFilesByGlob`, we need `obj.supportFile`. But the problem is that global `supportFile` option is legacy. It is splitted and moved under `e2e` or `component` from 10. 

It means that current `setSupportFileAndFolder` can be buggy or completely removable in the current version. I'm not 100% sure about this. I need to think about a test case that shows this problem. Maybe #22776 is the related case.

The goal of this PR is copying and moving `server/lib/config.ts` to `pkg/config` and adding types as much as possible. I focused on this because this PR can be too big and it makes reviewers hard to review this PR. 

I think checking/fixing bugs and adding more specific types should be done in the future PRs. What do you think?

Reply:
> To call the ctx.file.getFilesByGlob, we need obj.supportFile. But the problem is that global supportFile option is legacy. It is splitted and moved under e2e or component from 10.

This log is correct, at this point we are testing the flattened configuration for the given testing type, so `obj.supportFile` is valid.

I think making a few small changes seems needed to move this code over and would be worth making a small change to remove this circular dependency. Simply passing int he attribute needed is probably the simplest approach until / if we ever want to revisit how this is composed.

```js
// data-context/ProjectConfigManger.ts
setupFullConfigWithDefaults({
  projectRoot,
  projectName,
  configFile,
  config,
  envFile,
  options,
  cliConfig
}, 
ctx.file.getFilesByGlob)
```

Reply:
I never thought about passing the function as an argument to remove the dependency. Thanks for the idea. 

Unfortunately, this idea even cannot completely remove `pkg/data-context` from `config`. It can be removed from `src`. But it had to be added to `test` to run tests. 

However, I think this change is much better.